spark/src/main/scala/org/apache/spark/sql/delta/commands/UpdateCommand.scala

```diff
@@ -241,35 +319,60 @@ case class UpdateCommand(
    * When CDF is enabled, includes the generation of CDC preimage and postimage columns for
    * changed rows.
    *
-   * @return the list of [[AddFile]]s and [[AddCDCFile]]s that have been written.
+   * @return a list of [[FileAction]]s, consisting of newly-written data and CDC files and old
+   *         files that have been removed.
    */
   private def rewriteFiles(
       spark: SparkSession,
       txn: OptimisticTransaction,
       rootPath: Path,
-      inputLeafFiles: Seq[String],
+      inputLeafFiles: Seq[AddFile],
       nameToAddFileMap: Map[String, AddFile],
-      condition: Expression): Seq[FileAction] = {
+      condition: Expression,
+      generateRemoveFileActions: Boolean,
+      copyUnmodifiedRows: Boolean): Seq[FileAction] = {
+    // Number of total rows that we have seen, i.e. are either copying or updating (sum of both).
+    // This will be used later, along with numUpdatedRows, to determine numCopiedRows.
+    val incrTouchedCountExpr = IncrementMetric(TrueLiteral, metrics("numTouchedRows"))
+
     // Containing the map from the relative file path to AddFile
     val baseRelation = buildBaseRelation(
-      spark, txn, "update", rootPath, inputLeafFiles, nameToAddFileMap)
+      spark, txn, "update", rootPath, inputLeafFiles.map(_.path), nameToAddFileMap)
     val newTarget = DeltaTableUtils.replaceFileIndex(target, baseRelation.location)
     val targetDf = Dataset.ofRows(spark, newTarget)
-
-    // Number of total rows that we have seen, i.e. are either copying or updating (sum of both).
-    // This will be used later, along with numUpdatedRows, to determine numCopiedRows.
-    val incrTouchedCountExpr = IncrementMetric(TrueLiteral, metrics("numTouchedRows"))
+    val targetDfWithEvaluatedCondition = {
+      val evalDf = targetDf.withColumn(UpdateCommand.CONDITION_COLUMN_NAME, new Column(condition))
+      val copyAndUpdateRowsDf = if (copyUnmodifiedRows) {
+        evalDf
+      } else {
+        evalDf.filter(new Column(UpdateCommand.CONDITION_COLUMN_NAME))
+      }
+      copyAndUpdateRowsDf.filter(new Column(incrTouchedCountExpr))
+    }
 
     val updatedDataFrame = UpdateCommand.withUpdatedColumns(
       target.output,
       updateExpressions,
       condition,
-      targetDf
-        .filter(new Column(incrTouchedCountExpr))
-        .withColumn(UpdateCommand.CONDITION_COLUMN_NAME, new Column(condition)),
+      targetDfWithEvaluatedCondition,
       UpdateCommand.shouldOutputCdc(txn))
 
-    txn.writeFiles(updatedDataFrame)
+    val addFiles = txn.writeFiles(updatedDataFrame)
+
+    val removeFiles = if (generateRemoveFileActions) {
```

Comment:
Aren't remove actions being generated by `processUnmodifiedData`? Why do we need this one?